### DISCUSSION TITLE
[v2] stuck on isLoading when I setQueryData for initial load

### DISCUSSION THREAD

**Discussion by huchenme** [TanStack/query#649] [General] [answered]
**Describe the bug**

For my app, as the initial data was saved in local storage, I do not wish to make network call and show spinner. After upgrade to v2, I noticed that the app stuck at `isLoading` equal to true
https://codesandbox.io/s/react-query-2-setquerydata-18k47

**To Reproduce**
1. Go to https://codesandbox.io/s/react-query-2-setquerydata-18k47
2. See the screen stuck at "Loading..." (you might need to click the browser refresh on the right panel)

**Expected behavior**
It should load the data in `queryCache`

**Screenshots**
![image](https://user-images.githubusercontent.com/2078389/85949842-9df6a080-b98b-11ea-88c3-e2e8eda0cb41.png)

**Desktop (please complete the following information):**
 - OS: [e.g. iOS]: macOs catalina
 - Browser [e.g. chrome, safari]: chrome
 - Version [e.g. 22]: 83

**Additional context**
I tried another approach using `initialData`, but it is not working as expected as well: https://github.com/tannerlinsley/react-query/issues/650

**Comment by tannerlinsley** [comment]
I think this is what you're looking to do: https://codesandbox.io/s/react-query-2-setquerydata-5765w?file=/src/index.js

**Reply by huchenme** [reply]
Thanks, this solves my issue. But do you think for the `setQueryData` example I provided, it is a bug?

**Reply by tannerlinsley** [reply]
How so?

**Reply by huchenme** [reply]
as in the codesandbox link, it stuck in `loading...`, which I believe should not be expected for all edge cases like this

**Comment by tannerlinsley** [comment, marked as answer]
Pretty sure the reason for this is:
- You are attempting to set the query data *after* the query has started to load. Setting the query data with `setQueryData` will not bring a query out of a loading mode.
- Instead of using the `initialData` as is prescribed and designated in the docs, you are using setQueryData imperatively. The only way to use `setQueryData` and avoid the loading state is to ensure that it is done *before* the query for it loads.

https://codesandbox.io/s/react-query-2-setquerydata-9rm0m?file=/src/index.js
